core/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -796,25 +777,25 @@ public ClusterStateDiff(ClusterState before, ClusterState after) {
             customs = DiffableUtils.diff(before.customs, after.customs, DiffableUtils.getStringKeySerializer());
         }
 
-        public ClusterStateDiff(StreamInput in, ClusterState proto) throws IOException {
+        public ClusterStateDiff(StreamInput in, ClusterState proto, CustomPrototypeRegistry registry) throws IOException {
             clusterName = new ClusterName(in);
             fromUuid = in.readString();
             toUuid = in.readString();
             toVersion = in.readLong();
-            routingTable = proto.routingTable.readDiffFrom(in);
-            nodes = proto.nodes.readDiffFrom(in);
-            metaData = proto.metaData.readDiffFrom(in);
-            blocks = proto.blocks.readDiffFrom(in);
+            routingTable = proto.routingTable.readDiffFrom(in, registry);
+            nodes = proto.nodes.readDiffFrom(in, registry);
+            metaData = proto.metaData.readDiffFrom(in, registry);
+            blocks = proto.blocks.readDiffFrom(in, registry);
             customs = DiffableUtils.readImmutableOpenMapDiff(in, DiffableUtils.getStringKeySerializer(),
                 new DiffableUtils.DiffableValueSerializer<String, Custom>() {
                     @Override
                     public Custom read(StreamInput in, String key) throws IOException {
-                        return lookupPrototypeSafe(key).readFrom(in);
+                        return registry.getClusterStatePrototype(key).readFrom(in);
```

Comment:
This is for bwc reasons. The `key` already represents the name of the custom metadata. If we would use the `readNamedWriteable(...)` here than we would read the name of the custom metadata twice. If not for bwc this `read(StreamInput, String)` method should be refactored. Shall I leave a TODO here?